CI: Travis - TestGetIdentity/Duplicated_key timed out

## CI failure

Hit on https://github.com/cilium/cilium/pull/29983

```
[FAIL]	   600.04s	github.com/cilium/cilium/pkg/k8s/identitybackend
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓                                                          
┃   PANIC  package: github.com/cilium/cilium/pkg/k8s/identitybackend • TestGetIdentity/Duplicated_key   ┃                                                          
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛                                                          
panic: test timed out after 10m0s                                                                                                                                  
running tests:                                                                                                                                                     
	TestGetIdentity (10m0s)                                                                                                                                            
	TestGetIdentity/Duplicated_key (10m0s)                                                                                                                             
                                                                                                                                                                   
goroutine 50 [running]:                                                                                                                                            
testing.(*M).startAlarm.func1()                                                                                                                                    
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:2259 +0x320                                                                               
created by time.goFunc                                                                                                                                             
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/time/sleep.go:176 +0x38                                                                                      
                                                                                                                                                                   
goroutine 1 [chan receive, 10 minutes]:                                                                                                                            
testing.(*T).Run(0x40003ffba0, {0x1fa481b?, 0x1601d68832e?}, 0x20c7a20)                                                                                            
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:1649 +0x350                                                                               
testing.runTests.func1(0x400051ab40?)                                                                                                                              
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:2054 +0x48                                                                                
testing.tRunner(0x40003ffba0, 0x40000cdc28)                                                                                                                        
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:1595 +0xe8                                                                                
testing.runTests(0x4000471a40?, {0x3620280, 0x2, 0x2}, {0x40?, 0x1d8bae0?, 0x3650fa0?})                                                                            
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:2052 +0x3b4                                                                               
testing.(*M).Run(0x4000471a40)                                                                                                                                     
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:1925 +0x538                                                                               
main.main()                                                                                                                                                        
	_testmain.go:83 +0x1c8                                                                                                                                             
                                                                                                                                                                   
goroutine 25 [chan receive, 10 minutes]:                                                                                                                           
testing.(*T).Run(0x400025a000, {0x1fa10b5?, 0x0?}, 0x4000551710)                                                                                                   
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:1649 +0x350                                                                               
github.com/cilium/cilium/pkg/k8s/identitybackend.TestGetIdentity(0x400025a000?)                                                                                    
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/identitybackend/identity_test.go:204 +0x940                                                               
testing.tRunner(0x400025a000, 0x20c7a20)                                                                                                                           
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:1595 +0xe8                                                                                
created by testing.(*T).Run in goroutine 1                                                                                                                         
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:1648 +0x33c                                                                               
                                                                                                                                                                   
goroutine 42 [semacquire, 10 minutes]:                                                                                                                             
sync.runtime_Semacquire(0x22e00c8?)                                                                                                                                
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/runtime/sema.go:62 +0x2c                                                                                     
sync.(*WaitGroup).Wait(0x40001448f0)                                                                                                                               
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/sync/waitgroup.go:116 +0x74                                                                                  
github.com/cilium/cilium/pkg/k8s/identitybackend.TestGetIdentity.func1(0x400025ba00)                                                                               
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/identitybackend/identity_test.go:241 +0x4cc                                                               
testing.tRunner(0x400025ba00, 0x4000551710)                                                                                                                        
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:1595 +0xe8                                                                                
created by testing.(*T).Run in goroutine 25                                                                                                                        
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/testing/testing.go:1648 +0x33c                                                                               
                                                                                                                                                                   
goroutine 27 [select]:                                                                                                                                             
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x40001309c0?, {0x22d5b60, 0x40006fe000}, 0x1, 0x400005b8c0)                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:238 +0xf8                                                     
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x4000144490?, 0x3b9aca00, 0x0, 0xc0?, 0xffff77561c58?)                                                              
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x80                                                     
k8s.io/apimachinery/pkg/util/wait.Until(...)                                                                                                                       
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161                                                           
k8s.io/client-go/tools/cache.(*controller).Run(0x40004e0960, 0x400005b8c0)                                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:159 +0x2d0                                                      
github.com/cilium/cilium/pkg/k8s/informer.(*privateRunner).Run(0x40004431a0, 0x400005b8c0)                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:46 +0xa8                                                                             
github.com/cilium/cilium/pkg/k8s/identitybackend.(*crdBackend).ListAndWatch(0x400051be90, {0x0?, 0x0?}, {0x22fc280?, 0x4000550790}, 0x400005b8c0)                  
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/identitybackend/identity.go:407 +0x540                                                                    
created by github.com/cilium/cilium/pkg/k8s/identitybackend.TestGetIdentity.func1 in goroutine 26                                                                  
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/identitybackend/identity_test.go:225 +0x2e4                                                               
                                                                                                                                                                   
goroutine 7 [sync.Cond.Wait, 10 minutes]:                                                                                                                          
sync.runtime_notifyListWait(0x40004f1998, 0x1)                                                                                                                     
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/runtime/sema.go:527 +0x154                                                                                   
sync.(*Cond).Wait(0x40004f1988)                                                                                                                                    
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/sync/cond.go:70 +0xcc                                                                                        
k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0x40004f1970, 0x40000e00c0)                                                                                          
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:575 +0x1dc                                                      
k8s.io/client-go/tools/cache.(*controller).processLoop(0x40004fa780)                                                                                               
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:188 +0x3c                                                       
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x40000da270?)                                                                                                
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x40                                                     
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x400025a340?, {0x22d5b60, 0x40000d8660}, 0x1, 0x4000100300)                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0x90                                                     
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x40002c20d0?, 0x3b9aca00, 0x0, 0xc0?, 0xffff77561c58?)                                                              
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x80                                                     
k8s.io/apimachinery/pkg/util/wait.Until(...)                                                                                                                       
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161                                                           
k8s.io/client-go/tools/cache.(*controller).Run(0x40004fa780, 0x4000100300)                                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:159 +0x2d0                                                      
github.com/cilium/cilium/pkg/k8s/informer.(*privateRunner).Run(0x40000f0100, 0x4000100300)                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:46 +0xa8                                                                             
github.com/cilium/cilium/pkg/k8s/identitybackend.(*crdBackend).ListAndWatch(0x40000d8510, {0x0?, 0x0?}, {0x22fc280?, 0x400037a640}, 0x4000100300)                  
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/identitybackend/identity.go:407 +0x540                                                                    
created by github.com/cilium/cilium/pkg/k8s/identitybackend.TestGetIdentity.func1 in goroutine 6                                                                   
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/identitybackend/identity_test.go:225 +0x2e4                                                               
                                                                                                                                                                   
goroutine 31 [select, 10 minutes]:                                                                                                                                 
k8s.io/client-go/tools/cache.watchHandler({0x0?, 0x0?, 0x3650fa0?}, {0x22e5278, 0x400043b1d0}, {0xffff77561c58, 0x4000243e40}, {0x231cde8?, 0x1f2f3e0}, 0x0, ...)  
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:724 +0x120                                                       
k8s.io/client-go/tools/cache.(*Reflector).watch(0x400041cb60, {0x0?, 0x0?}, 0x400005b8c0, 0x0?)                                                                    
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:437 +0x3e0                                                       
k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0x400041cb60, 0x400005b8c0)                                                                                 
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:357 +0x3f4                                                       
k8s.io/client-go/tools/cache.(*Reflector).Run.func1()                                                                                                              
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:291 +0x28                                                        
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x0?)                                                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x40                                                     
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x40003cef80?, {0x22d5b80, 0x40006fc000}, 0x1, 0x400005b8c0)                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0x90                                                     
k8s.io/client-go/tools/cache.(*Reflector).Run(0x400041cb60, 0x400005b8c0)                                                                                          
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:290 +0x184                                                       
k8s.io/client-go/tools/cache.(*controller).Run.(*Group).StartWithChannel.func2()                                                                                   
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:55 +0x2c                                                         
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()                                                                                                           
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x58                                                         
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 27                                                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x7c                                                         
                                                                                                                                                                   
goroutine 33 [select, 10 minutes]:                                                                                                                                 
k8s.io/client-go/tools/cache.(*Reflector).startResync(0x400041cb60, 0x400005b8c0, 0x400005b9e0, 0x4000077f08?)                                                     
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:369 +0xbc                                                        
created by k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch in goroutine 31                                                                                  
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:356 +0x3dc                                                       
                                                                                                                                                                   
goroutine 35 [select]:                                                                                                                                             
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x40000fa1a0?, {0x22d5b60, 0x40006feb10}, 0x1, 0x400005baa0)                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:238 +0xf8                                                     
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x4000144670?, 0x3b9aca00, 0x0, 0xe0?, 0xffff77561c58?)                                                              
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x80                                                     
k8s.io/apimachinery/pkg/util/wait.Until(...)                                                                                                                       
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161                                                           
k8s.io/client-go/tools/cache.(*controller).Run(0x40004e14a0, 0x400005baa0)                                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:159 +0x2d0                                                      
github.com/cilium/cilium/pkg/k8s/informer.(*privateRunner).Run(0x40004434e0, 0x400005baa0)                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:46 +0xa8                                                                             
github.com/cilium/cilium/pkg/k8s/identitybackend.(*crdBackend).ListAndWatch(0x40006fe9c0, {0x400037a640?, 0x400007dfb0?}, {0x22fc280?, 0x4000551260}, 0x400005baa0)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/identitybackend/identity.go:407 +0x540                                                                    
created by github.com/cilium/cilium/pkg/k8s/identitybackend.TestGetIdentity.func1 in goroutine 34                                                                  
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/identitybackend/identity_test.go:225 +0x2e4                                                               
                                                                                                                                                                   
goroutine 10 [chan receive, 10 minutes]:                                                                                                                           
k8s.io/client-go/tools/cache.(*controller).Run.func1()                                                                                                             
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:132 +0x2c                                                       
created by k8s.io/client-go/tools/cache.(*controller).Run in goroutine 7                                                                                           
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:131 +0x94                                                       
                                                                                                                                                                   
goroutine 11 [select, 10 minutes]:                                                                                                                                 
k8s.io/client-go/tools/cache.watchHandler({0x0?, 0x0?, 0x3650fa0?}, {0x22e5278, 0x40000da3d8}, {0xffff77561c58, 0x40004f1970}, {0x231cde8?, 0x1f2f3e0}, 0x0, ...)  
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:724 +0x120                                                       
k8s.io/client-go/tools/cache.(*Reflector).watch(0x40000fc0e0, {0x0?, 0x0?}, 0x4000100300, 0x0?)                                                                    
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:437 +0x3e0                                                       
k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0x40000fc0e0, 0x4000100300)                                                                                 
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:357 +0x3f4                                                       
k8s.io/client-go/tools/cache.(*Reflector).Run.func1()                                                                                                              
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:291 +0x28                                                        
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x0?)                                                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x40                                                     
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x40000ec040?, {0x22d5b80, 0x40000d6230}, 0x1, 0x4000100300)                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0x90                                                     
k8s.io/client-go/tools/cache.(*Reflector).Run(0x40000fc0e0, 0x4000100300)                                                                                          
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:290 +0x184                                                       
k8s.io/client-go/tools/cache.(*controller).Run.(*Group).StartWithChannel.func2()                                                                                   
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:55 +0x2c                                                         
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()                                                                                                           
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x58                                                         
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 7                                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x7c                                                         
                                                                                                                                                                   
goroutine 39 [select, 10 minutes]:                                                                                                                                 
k8s.io/client-go/tools/cache.watchHandler({0x0?, 0x0?, 0x3650fa0?}, {0x22e5278, 0x400043b668}, {0xffff77561c58, 0x40007782c0}, {0x231cde8?, 0x1f2f3e0}, 0x0, ...)  
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:724 +0x120                                                       
k8s.io/client-go/tools/cache.(*Reflector).watch(0x400041cfc0, {0x0?, 0x0?}, 0x400005baa0, 0x0?)                                                                    
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:437 +0x3e0                                                       
k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0x400041cfc0, 0x400005baa0)                                                                                 
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:357 +0x3f4                                                       
k8s.io/client-go/tools/cache.(*Reflector).Run.func1()                                                                                                              
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:291 +0x28                                                        
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x0?)                                                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x40                                                     
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x40003cefc0?, {0x22d5b80, 0x40006fc1e0}, 0x1, 0x400005baa0)                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0x90                                                     
k8s.io/client-go/tools/cache.(*Reflector).Run(0x400041cfc0, 0x400005baa0)                                                                                          
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:290 +0x184                                                       
k8s.io/client-go/tools/cache.(*controller).Run.(*Group).StartWithChannel.func2()                                                                                   
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:55 +0x2c                                                         
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()                                                                                                           
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x58                                                         
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 35                                                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x7c                                                         
                                                                                                                                                                   
goroutine 41 [select, 10 minutes]:                                                                                                                                 
k8s.io/client-go/tools/cache.(*Reflector).startResync(0x400041cfc0, 0x400005baa0, 0x400005bbc0, 0x4000079708?)                                                     
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:369 +0xbc                                                        
created by k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch in goroutine 39                                                                                  
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:356 +0x3dc                                                       
                                                                                                                                                                   
goroutine 43 [sync.Cond.Wait, 10 minutes]:                                                                                                                         
sync.runtime_notifyListWait(0x4000778658, 0x0)                                                                                                                     
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/runtime/sema.go:527 +0x154                                                                                   
sync.(*Cond).Wait(0x4000778648)                                                                                                                                    
	/home/travis/.gimme/versions/go1.21.6.linux.arm64/src/sync/cond.go:70 +0xcc                                                                                        
k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0x4000778630, 0x4000441440)                                                                                          
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:575 +0x1dc                                                      
k8s.io/client-go/tools/cache.(*controller).processLoop(0x4000471f40)                                                                                               
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:188 +0x3c                                                       
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x400043b938?)                                                                                                
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x40                                                     
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x400025bd40?, {0x22d5b60, 0x40006ff3e0}, 0x1, 0x400005bc80)                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0x90                                                     
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x4000144910?, 0x3b9aca00, 0x0, 0x60?, 0xffff77561c58?)                                                              
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x80                                                     
k8s.io/apimachinery/pkg/util/wait.Until(...)                                                                                                                       
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161                                                           
k8s.io/client-go/tools/cache.(*controller).Run(0x4000471f40, 0x400005bc80)                                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:159 +0x2d0                                                      
github.com/cilium/cilium/pkg/k8s/informer.(*privateRunner).Run(0x40004437c0, 0x400005bc80)                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:46 +0xa8                                                                             
github.com/cilium/cilium/pkg/k8s/identitybackend.(*crdBackend).ListAndWatch(0x40006ff290, {0x0?, 0x0?}, {0x22fc280?, 0x4000551860}, 0x400005bc80)                  
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/identitybackend/identity.go:407 +0x540                                                                    
created by github.com/cilium/cilium/pkg/k8s/identitybackend.TestGetIdentity.func1 in goroutine 42                                                                  
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/identitybackend/identity_test.go:225 +0x2e4                                                               
                                                                                                                                                                   
goroutine 46 [chan receive, 10 minutes]:                                                                                                                           
k8s.io/client-go/tools/cache.(*controller).Run.func1()                                                                                                             
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:132 +0x2c                                                       
created by k8s.io/client-go/tools/cache.(*controller).Run in goroutine 43                                                                                          
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:131 +0x94                                                       
                                                                                                                                                                   
goroutine 47 [select, 10 minutes]:                                                                                                                                 
k8s.io/client-go/tools/cache.watchHandler({0x0?, 0x0?, 0x3650fa0?}, {0x22e5278, 0x40000da438}, {0xffff77561c58, 0x4000778630}, {0x231cde8?, 0x1f2f3e0}, 0x0, ...)  
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:724 +0x120                                                       
k8s.io/client-go/tools/cache.(*Reflector).watch(0x400041d420, {0x0?, 0x0?}, 0x400005bc80, 0x0?)                                                                    
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:437 +0x3e0                                                       
k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0x400041d420, 0x400005bc80)                                                                                 
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:357 +0x3f4                                                       
k8s.io/client-go/tools/cache.(*Reflector).Run.func1()                                                                                                              
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:291 +0x28                                                        
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x0?)                                                                                                         
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x40                                                     
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x40003cf000?, {0x22d5b80, 0x40006fc3c0}, 0x1, 0x400005bc80)                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0x90                                                     
k8s.io/client-go/tools/cache.(*Reflector).Run(0x400041d420, 0x400005bc80)                                                                                          
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:290 +0x184                                                       
k8s.io/client-go/tools/cache.(*controller).Run.(*Group).StartWithChannel.func2()                                                                                   
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:55 +0x2c                                                         
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()                                                                                                           
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x58                                                         
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 43                                                                                        
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x7c                                                         
                                                                                                                                                                   
goroutine 49 [select, 10 minutes]:                                                                                                                                 
k8s.io/client-go/tools/cache.(*Reflector).startResync(0x400041d420, 0x400005bc80, 0x400005bda0, 0x4000782708?)                                                     
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:369 +0xbc                                                        
created by k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch in goroutine 47                                                                                  
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/reflector.go:356 +0x3dc                                                       
FAIL	github.com/cilium/cilium/pkg/k8s/identitybackend	600.039s
```

--- a/pkg/k8s/identitybackend/identity_test.go
+++ b/pkg/k8s/identitybackend/identity_test.go
@@ -222,19 +222,26 @@ func TestGetIdentity(t *testing.T) {
 			addWaitGroup := sync.WaitGroup{}
 			addWaitGroup.Add(len(tc.identities))
 
+			// To avoid a race, we must create these before we start ListAndWatch, see #30873. There
+			// is no easy way of knowing when the watch is established. Specifically, 'HasSynced'
+			// does _not_ guarantee it: the fake object tracker doesn't do resource versioning and
+			// hence cannot replay events in the reflector's gap between list and watch. Ironically,
+			// therefore, if we waited for the informer's HasSynced, we'd _increase_ the likelihood
+			// of the race. Avoid the whole issue by creating the objects before the informer is
+			// even started, thus guaranteeing the objects are part of the initial list.
+			for _, identity := range tc.identities {
+				_, err = client.CiliumV2().CiliumIdentities().Create(ctx, &identity, v1.CreateOptions{})
+				if err != nil {
+					t.Fatalf("Can't create identity %s: %s", identity.Name, err)
+				}
+			}
+
 			go backend.ListAndWatch(ctx, FakeHandler{onListDoneChan: listenerReadyChan, onAddFunc: func() { addWaitGroup.Done() }}, stopChan)
 
 			select {
 			case <-listenerReadyChan:
 			case <-inctimer.After(2 * time.Second):
 				t.Fatalf("Failed to listen for identities within 2 seconds")
-			}
-
-			for _, identity := range tc.identities {
-				_, err = client.CiliumV2().CiliumIdentities().Create(ctx, &identity, v1.CreateOptions{})
-				if err != nil {
-					t.Fatalf("Can't create identity %s: %s", identity.Name, err)
-				}
 			}
 
 			// Wait for watcher to process the identities in the background

--- a/pkg/k8s/identitybackend/identity_test.go
+++ b/pkg/k8s/identitybackend/identity_test.go
@@ -18,7 +18,6 @@ import (
 	"github.com/cilium/cilium/pkg/checker"
 	"github.com/cilium/cilium/pkg/identity/key"
 	"github.com/cilium/cilium/pkg/idpool"
-	"github.com/cilium/cilium/pkg/inctimer"
 	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
 	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
 	"github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1/validation"
@@ -96,15 +95,10 @@ func (s *K8sIdentityBackendSuite) TestSanitizeK8sLabels(c *C) {
 }
 
 type FakeHandler struct {
-	onListDoneChan chan struct{}
-	onAddFunc      func()
+	onAddFunc func()
 }
 
-func (f FakeHandler) OnListDone() {
-	if f.onListDoneChan != nil {
-		close(f.onListDoneChan)
-	}
-}
+func (f FakeHandler) OnListDone() {}
 
 func (f FakeHandler) OnAdd(id idpool.ID, key allocator.AllocatorKey) {
 	if f.onAddFunc != nil {
@@ -201,7 +195,9 @@ func TestGetIdentity(t *testing.T) {
 	}
 
 	for _, tc := range testCases {
+		tc := tc
 		t.Run(tc.desc, func(t *testing.T) {
+			t.Parallel()
 			_, client := k8sClient.NewFakeClientset()
 			backend, err := NewCRDBackend(CRDBackendConfiguration{
 				Store:   nil,
@@ -213,10 +209,9 @@ func TestGetIdentity(t *testing.T) {
 			}
 
 			ctx := context.Background()
-			stopChan := make(chan struct{}, 1)
-			listenerReadyChan := make(chan struct{}, 1)
+			stopChan := make(chan struct{})
 			defer func() {
-				stopChan <- struct{}{}
+				close(stopChan)
 			}()
 
 			addWaitGroup := sync.WaitGroup{}
@@ -236,13 +231,7 @@ func TestGetIdentity(t *testing.T) {
 				}
 			}
 
-			go backend.ListAndWatch(ctx, FakeHandler{onListDoneChan: listenerReadyChan, onAddFunc: func() { addWaitGroup.Done() }}, stopChan)
-
-			select {
-			case <-listenerReadyChan:
-			case <-inctimer.After(2 * time.Second):
-				t.Fatalf("Failed to listen for identities within 2 seconds")
-			}
+			go backend.ListAndWatch(ctx, FakeHandler{onAddFunc: func() { addWaitGroup.Done() }}, stopChan)
 
 			// Wait for watcher to process the identities in the background
 			addWaitGroup.Wait()

